### ISSUE TITLE
Async HTML Imports breaks data-binding in certain instances

### ISSUE THREAD

**Issue by jshcrowthe** [Polymer/polymer#2678]
When using the `async` attribute on a HTML import the data-binding is inconsistent about when it will and will not work.

I have documented a simple instance I was able to reproduce. It happens when attributes are set via javascript as opposed to being declared in the HTML. 

http://plnkr.co/edit/tNUEQKv1REnqmigxdTxr

I have also in a local project reproduced this for data-binding between elements but am still working on reproducing it in a plunkr.

**Comment by ciampo** [comment]
Works if you change ```el.name = 'John Doe';``` to ```el.setAttribute('name', 'John Doe');```

**Comment by ebidel** [comment]
This is because in your example, you've changed the timing of upgrading simple-component. Namely, you're setting it's `.name` property before the custom element has been registered and has a .name property.

That said, for an example like this, you should always wait for the `WebComponentsReady` event to fire before setting a property synchronously. For example, if you remove `async` from the import, I suspect your code will fail in Safari/FF/Edge.

http://webcomponents.org/polyfills/html-imports/

**Comment by ebidel** [comment]
Here's an example: http://plnkr.co/edit/nTIRCpZZewSp6R6oAB4e?p=preview

**Comment by jshcrowthe** [comment]
So the event works great when loading the Polyfills on all browsers seeing as the `WebComponentsReady` event and `HTMLImportsLoaded` events are fired by the polyfill itself. However if I am conditionally loading the Polyfill and that event doesn't get fired, then the issue remains.

I guess I could manually fire a `WebComponentsReady` (or for your example `HTMLImportsLoaded`) when my element(s) are loaded and then I could reliably key off of that event. But to me that doesn't seem ideal.

**Comment by ebidel** [comment]
There's more to this story :) So if you're conditionally loading the polyfills than yes, this event may not fire. 
In that case, you could use an `ononload` handler on the async import.

**Comment by jshcrowthe** [comment]
Indeed :) Alright I can live with that but it might be nice if Polymer was to try and grab those properties of the node itself instead of just the attributes (although despite my desires for things to just do what I tell it, it does make sense that it works that way).

I am working on getting you a plunkr of a broken data-binding example. Unfortunately for me the only reproducible example I have is behind a layer of authentication at the moment (lame I know). It is blowing my mind why it's not working so I will try to get you an example in the next day or two.

**Comment by ebidel** [comment]
Just note that `el.setAttribute` works in this case because the value is a string. If `name` was an object/array, `setAttribute` wouldn't work because it only works with attributes (html strings).

**Comment by jshcrowthe** [comment]
Gotcha! Well in that case I will work on that Plunkr and hopefully get it to you soon so you can see what error I'm having!

**Comment by jshcrowthe** [comment]
Alright, I spun up a dummy API with some dummy data that illustrates my issue (it's hosted on a heroku free dyno so if the first time you hit the page the API request may take forever). But the data-binding issue is visible for you to look at:

http://jshcrowthe.github.io/gallery-data/demo/

Let me know if I'm just doing something weird (I love it when there is a lesson to be learned!). As with the previous plunkr you can mess with the async tag on line 17 to see the working and non-working state:

https://github.com/jshcrowthe/gallery-data/blob/gh-pages/demo/index.html

I stripped out some things (somewhat haphazardly) to try and isolate things a little more. Pardon the commit history on the `gh-pages` branch.

**Comment by jshcrowthe** [comment]
@ebidel: Any news on this?

**Comment by ebidel** [comment]
I believe this is related to https://github.com/Polymer/polymer/issues/2481, which ultimately a bug we need to file against chrome. When you go `async` on your main page import, the subimports are loaded out of order. What I've seen happen is that the order of [these imports](https://github.com/jshcrowthe/gallery-data/blob/gh-pages/demo/elements.html) and make gallery-data.html last, I suspect it may work. That's what I've experienced at least.

**Comment by jshcrowthe** [comment]
I went and tested the demo I provided in Firefox and Safari and both of those seemed to work just fine. So I decided to tinker with what you'd said (in addition to some of the comments you'd made on other issues (Specifically the vulcanized import suggestion from #2522) and I think I may have better reproduced the specific issue I'm having (In addition to potentially offering some more information).

As you mentioned moving `gallery-data.html` to my final import and that seemed to fix the error for this specific case.

If I vulcanized my `elements.html` file into `bundle.html` that I then included instead of the `elements.html` file. This fixed the issue and my element bundle functions as it should (regardless of the position of `gallery-data.html` pre-bundling, but that's a no brainer seeing as it gets compiled into one file).

After that I conditionally injected the file `webcomponents-lite.min.js` and once again my page breaks (even with a vulcanized bundle) leaving me back where I started.

Firefox and Safari (I'm assuming because they load the Polyfills) work fine throughout this whole process. 

Hope this is useful, would love to see this get reopened for some more investigation.
